### DISCUSSION TITLE
{} Initializer for const structs

### DISCUSSION THREAD

**Discussion by adaumann** [drmortalwombat/oscar64#174] [Q&A]
I would like to init const lists of structs that should be loaded into an easyflash bank.The {} initalizer from C++ are not working. How to implement this? Or is it planned to support it for Oscar64

Example:
Creature_t creatures_list[MON_MAX_CREATURES] = {
    {"Filthy Street Urchin",      0x0012000AL, 0x00000000L, 0x2034,     0,  40,  4,   1, 11, 'p', {  1,  4}, { 72, 148,   0,   0},   0},

The struct looks like (has sub structures):

typedef struct Creature_t {
    const char *name;           // Description of creature
    uint32_t movement;          // Bit field
    uint32_t spells;            // Creature spells
    uint16_t defenses;          // Bit field
    uint16_t kill_exp_value;    // Exp value for kill
    uint8_t sleep_counter;      // Inactive counter / 10
    uint8_t area_affect_radius; // Area affect radius
    uint8_t ac;                 // AC
    uint8_t speed;              // Movement speed+10 (NOTE: +10 so that it can be an unsigned int)
    uint8_t sprite;             // Character representation (cchar)
    Dice_t hit_die;             // Creatures hit die
    uint8_t damage[4];          // Type attack and damage
    uint8_t level;              // Level of creature
} Creature_t; 
   
    
Andy

**Comment by drmortalwombat** [comment]
I don't see the issue, can you give a more complete example please?

**Reply by adaumann** [reply]
I need to initialize a list of structs:

Creature_t creatures_list[MON_MAX_CREATURES];

The list is not changeable and I need to initialize it, normally I do it with {} initializers like 

Creature_t creatures_list[MON_MAX_CREATURES] = {
{"Filthy Street Urchin", 0x0012000AL, 0x00000000L, 0x2034, 0, 40, 4, 1, 11, 'p', { 1, 4}, { 72, 148, 0, 0}, 0},
{"Kobolt", 0x0012000AL, 0x00000000L, 0x2034, 0, 40, 4, 1, 11, 'p', { 1, 4}, { 72, 148, 0, 0}, 0}
};

The list should be placed in a easyflash bank as immutable list during compile time.

Alternatively I can write a function like "InitCreatures" and place the list in bank 0 but this will use extra memory that I  don't have.

The main problem is that OscarC is not supporting {} initializers at all like

Dice_t dice = Dice_t{1,6};

This would be very helpful, or I need an alternative

**Comment by drmortalwombat** [comment]
Oscar has C style {} initializers and C++ style () constructors, which is sufficient in most case, please provide a complete compilable example, so I can have a look what can be done.

**Reply by adaumann** [reply]
Ok, I will make an example...

**Reply by drmortalwombat** [reply]
Thank you

**Reply by adaumann** [reply]
Mmmh,

in my example it works with the init of structures, just the direct const will not work:

'''
typedef struct Dice_t {
    char dice;
    char sides;
} Dice_t;

Dice_t dices[2] = {
    {1,6},
    {1,12},
};

int diceRoll(Dice_t const &dice) {
    return 0;
}

int main(void)
{
   
    diceRoll(dices[0]);
    
    // not working:
    diceRoll(Dice_t{1,6});
	return 0;
}
'''

**Reply by adaumann** [reply]
Must be something different... I will do it later

**Reply by drmortalwombat** [reply]
You can try the C style initializer: 
`diceRoll((Dice_t){1,6});`

**Reply by drmortalwombat** [reply]
see [compound literal](https://en.cppreference.com/w/c/language/compound_literal)

**Comment by adaumann** [comment]
Was my mistake, you can close it
